Introduce custom ESLint rules to apply and enforce new themed component selector convention

ybnd opened the pull request:
## References
_Add references/links to any related issues or PRs. These may include:_
* Fixes #2845

## Description
This PR introduces **custom ESLint plugins** as a means of automatically migrating to the new themed component selector convention proposed in #2845, as well as enforcing this convention moving forward.

Furthermore, these plugins can be leveraged in future PRs to create more rules for DSpace Angular in particular.

Note that this PR also enables linting for our Cypress tests, since they are affected by the selector change (most of the changes there are indentation/import order).

## Instructions for Reviewers
- The first commit of this PR includes the bulk of the new rules, enough to cover the majority of cases
    - It feels right to bundle these plugins in the project itself, since they are tightly coupled to DSpace Angular. 
        - **Do you agree? Any comments on the structure?**
    - The new plugins do not add more than a few seconds to the total linting time.
        - You can re-confirm this by running `TIMING=1 yarn run eslint . --quiet`
        - The `dspace-angular-ts/themed-component-usages` rule is the most impactful -- it may be useful to optimize it further, but it won't have as much as an effect as #2864
    - Please observe the lint output in the GitHub build logs, or locally if you check out this commit. 
    **Are the messages sufficiently clear?**
- The "auto-migrate" commit includes _only_ the results of `yarn lint --fix`. The manual interventions can be found in the subsequent commits
    - These commits can serve as an example of what downstream forks will have to do to migrate to the new convention (but likely a lot _less_ than was necessary in this PR)
    - **Is this reasonable? Do we need more detailed instructions?**
- The app should remain functionally the same
    - Run locally to confirm
    - Check a few themeable components to confirm that
        - The new convention is correctly reflected in the DOM
        - Everything still renders exactly as before
    - Switch between the `dspace` and `custom` themes
- If you're using Windows, I'd greatly appreciate if you check whether all new `yarn` scripts work there as well (they have already been tested on MacOS and Linux).


## Instructions for fork developers

After upgrading your fork to DSpace 8.0, run `yarn lint --quiet --fix`
- This should automatically migrate any custom themeable/themed components in your project
- Build/test/deploy your project and verify that everything still works
  - If there were major problems with the migration, they will come up as compilation errors. These can include
    - Out-of-sync inputs between the base themeable component and its `ThemedComponent` wrapper
    - ...
  - Please open an issue on GitHub if the cause of the problem is not clear


## Future work
I've already worked on similar ESLint rules in a separate project; if we decide to merge this PR it may be useful to port them to the "internal" lint plugin:
- Enforce correct theme declarations in entry component decorators such as `@listableObjectComponent`
- Enforce globally unique entry component decorators (to avoid silent overrides)
- Enforce import aliases (e.g. `of as observableOf`)

And moving forward there are many other DSpace-isms that we could write rules for
- Ensure that the inputs and outputs of `ThemedComponent` wrappers match the base component exactly
- Ensure that component overrides in themes do not introduce new inputs or outputs
- ...


## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [ESLint](https://eslint.org/) validation using `yarn lint`
- [x] My PR doesn't introduce circular dependencies (verified via `yarn check-circ-deps`)
- [ ] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

ybnd commented:
Still some e2e failures, but I'm pretty sure that those tests were already flaky

ybnd commented:
Note: mby far most of the conflicts will be due to the automatic lint fixes, so I'll address them in one go when the PR is approved.

Probably best to drop https://github.com/DSpace/dspace-angular/pull/2865/commits/3753b29a7313c676d4c9c720a3a7b5be0bfa087d, https://github.com/DSpace/dspace-angular/pull/2865/commits/d15327cd1301c04ee9f22868506ead34baed2574, https://github.com/DSpace/dspace-angular/pull/2865/commits/db7a371d332155f641ef04f02b8d80398484c7bc and re-do everything at the end of the branch for clarity.

tdonohue commented:
Hi @ybnd : This has conflicts because we just merged the Angular 16 upgrade in #2871 .  If you have time to resolve the conflicts, both @atarix83 and I were planning on trying this out next week _during testathon_.  We likely won't be able to review this prior to testathon though, unfortunately.

ybnd commented:
@tdonohue @atarix83 done.

There were only very minor conflicts; I'll keep an eye out in case there's more in the next few days

artlowel reviewed (approved):
Thanks @ybnd!

I verified that the code changes look good and everything still works. I also tried it on a theme for a different project, and everything was fixed automatically (although I did get some errors due to #2750 and #2343 I had to fix manually)

Currently it does have merge conflicts though

ybnd commented:
Fixed e2e test failure and tacked on a workaround for #2946

atarix83 reviewed (changes requested):
thanks @ybnd  for this PR.

I started the review and the first issue i've encountered is when i've opened the project with webstorm.
Basically the eslint plugin gives me an error because can't load the new custom eslint rules
![image](https://github.com/DSpace/dspace-angular/assets/2486489/7cfa94ca-f132-45ed-b597-1246b966b75b)

I tried to play with webstorm eslint plugin configurations, without any results.

The problem is that the IDE doesn't recognize any lint errors if can't load the configurations.

Can I ask you which IDE are you using ?

ybnd commented:
@atarix83 I'm using IntelliJ IDEA, so ESLint support is probably the same in WebStorm.
To use these plugins you'll need to run `yarn build:lint` first. 
I already included it in the `yarn lint` step, but it should really be documented more clearly...

![20240418-144303](https://github.com/DSpace/dspace-angular/assets/31547038/34903a6c-cd55-4f60-bbfa-49ea011e2aa9)
![20240418-144354](https://github.com/DSpace/dspace-angular/assets/31547038/2ee2ecaa-8557-4ee8-9e1b-a4d7dbc00c7f)

atarix83 reviewed (commented):
thanks @ybnd  i confirm running the command resolves the problem. I'm wondering if there is a way to avoid to rum this command manually so that other developers do not run into the same problem. I think we could add a post install command to run it?

![image](https://github.com/DSpace/dspace-angular/assets/2486489/ef6cf41f-9b26-4a18-b7e7-08a482fb7444)

@artlowel  @ybnd  what do you think?

atarix83 reviewed (approved):
thanks @ybnd 

I've verified that now IDE can detect lint issue properly.

I've also verified that application overall works good. 

There is only a problem with docker in the pipeline to be checked

tdonohue commented:
@ybnd : Per @atarix83 's note above, I looked into the Docker build issues in this PR.  It appears this PR is currently throwing a build error when it attempts to build `dspace-angular-dist`.  Here's the error:
```
#11 71.98 $ yarn build:lint
#11 72.14 yarn run v1.22.19
#11 72.17 $ rimraf 'lint/dist/**/*.js' 'lint/dist/**/*.js.map' && tsc -b lint/tsconfig.json
#11 72.41 error TS5083: Cannot read file '/app/lint/tsconfig.json'.
#11 72.42 error Command failed with exit code 1.
#11 72.42 info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

The Docker build uses `[src]/Dockerfile.dist` as shown here: https://github.com/DSpace/dspace-angular/blob/main/.github/workflows/docker.yml#L41-L52

It's unclear to me if that `Dockerfile.dist` requires updates, or if this indicates an unexpected bug/behavior in the build process with this PR.

ybnd commented:
@tdonohue @atarix83 I'm not super familiar with the Docker setup, but it seems like it involves a `yarn install` with only `package.json` & `yarn.json`, so it's failing during the new `postinstall` script.

Added a condition to check whether the lint plugin sources are present before trying to build, let's see if that's enough 🤞

artlowel reviewed (approved):
Thanks @ybnd that looks good!

@tdonohue it would be good if you could still run a sanity check to verify the build works as it should on windows, given the changes in https://github.com/DSpace/dspace-angular/pull/2865/commits/728b561bf4d9e3187484130777270c0c03fc8cb1

ybnd commented on package.json:
```diff
@@ -94,6 +100,8 @@
     "deepmerge": "^4.3.1",
     "ejs": "^3.1.9",
+    "eslint-plugin-dspace-angular-html": "link:./lint/dist/src/rules/html",
+    "eslint-plugin-dspace-angular-ts": "link:./lint/dist/src/rules/ts",
```
No reason AFAIK, I've moved them.

tdonohue reviewed (changes requested):
@ybnd : I've tested this today in production mode (`yarn build:prod; yarn serve:ssr`).  While it works for the `dspace` theme, the `custom` theme is broken.  Here's what I see 

![custom-theme](https://github.com/DSpace/dspace-angular/assets/483997/ace9ea11-8f1b-4bc8-b60c-133896d9408f)

There are a number of issues that are visible.  The header no longer works, no ability to login or browse the site, much of the homepage content no longer displays, etc.  Individual item pages are no better.

This doesn't appear to be a caching issue on my end, as after installing this PR I ran the following:
```
yarn clean
yarn install
yarn build:prod
yarn serve:ssr
```
I also attempted a hard reload in the browser in case this is a CSS issue.

I'll note that with the `custom` theme enabled, there are SSR errors that occur such as this one:
```
ERROR TypeError: Cannot read properties of undefined (reading 'onDestroy')
    at Object.ɵɵpipe (C:\dspace-angular\dist\server\main.js:1:4326830)
    at template (C:\dspace-angular\dist\server\1210.js:1:3854)
    at executeTemplate (C:\dspace-angular\dist\server\main.js:1:4182823)
    at renderView (C:\dspace-angular\dist\server\main.js:1:4198373)
    at renderComponent (C:\dspace-angular\dist\server\main.js:1:4198106)
    at renderChildComponents (C:\dspace-angular\dist\server\main.js:1:4198794)
    at renderView (C:\dspace-angular\dist\server\main.js:1:4198836)
    at ComponentFactory.create (C:\dspace-angular\dist\server\main.js:1:4235830)
    at R3ViewContainerRef.createComponent (C:\dspace-angular\dist\server\main.js:1:4239623)
    at Object.next (C:\dspace-angular\dist\server\main.js:1:2265629)
```

Finally, not all of the new `yarn` commands are working for me on Windows.  Here's what works and what doesn't:
* `yarn build:lint` - Succeeds
* `yarn test:lint` - FAILS.  I get two errors in the `dspace-angular\lint\test\fixture\src\test.ts` file.
    ``` 
    1) TypeScript rules themed-component-usages invalid disallow direct usages of base class
    Message:
    Error: Can't infer project-absolute TS/resource path from: C:\dspace-angular\lint\test\fixture\src\test.ts
    Occurred while linting C:\dspace-angular\lint\test\fixture\src\test.ts:2
    Rule: "themed-component-usages"

    2) TypeScript rules themed-component-usages invalid disallow direct usages of base class, keep other imports
    Message:
    Error: Can't infer project-absolute TS/resource path from: C:\dspace-angular\lint\test\fixture\src\test.ts
    Occurred while linting C:\dspace-angular\lint\test\fixture\src\test.ts:2
    Rule: "themed-component-usages"
    ```
*  `yarn test:lint:nobuild` - FAILS with the same two errors (obviously)
*  `yarn lint` - FAILS.  I get 97 lint errors in the `custom` theme.  All of them say something like "Unthemed version of themable component should have a selector starting with 'ds-base-'". Maybe this is why the "custom" theme is not working for me?
*  `yarn lint:nobuild` - FAILS (same as `yarn lint` obviously)
*  `yarn lint-fix` - Succeeds. It fixes all 97 errors above, but changes 97 files in the "custom" theme in the same way.  Here's an example of the `diff` change (changed `ds-themed-admin-sidebar` to `ds-base-themed-admin-sidebar`):
    ```
    --- a/src/themes/custom/app/admin/admin-sidebar/admin-sidebar.component.ts
    +++ b/src/themes/custom/app/admin/admin-sidebar/admin-sidebar.component.ts
    @@ -15,7 +15,7 @@ import { AdminSidebarComponent as BaseComponent } from 
    '../../../../../app/admin
      * Component representing the admin sidebar
      */
     @Component({
    -  selector: 'ds-themed-admin-sidebar',
    +  selector: 'ds-base-themed-admin-sidebar'
    ```
*  `yarn docs:lint` - Succeeds, but changes all line endings in the `/docs/lint/**` subdirectories to CRLF instead of LF.  (This is not a serious issue...it's something we could live with if we had to)

While the overall code looks reasonable to me, this doesn't seem to be fully functional (at least not on Windows).

ybnd commented:
@tdonohue I've addressed the issues with the new yarn scripts on Windows, so this should be ready to review again once https://github.com/DSpace/dspace-angular/issues/2977 is resolved

tdonohue commented:
@ybnd : This has minor conflicts caused by the fix to the `custom` theme: #2984   Once those conflicts are resolved, I should be able to test this again (hopefully on/around Monday)

ybnd commented:
@tdonohue couldn't make Monday, hopefully today is still ok timing-wise.
